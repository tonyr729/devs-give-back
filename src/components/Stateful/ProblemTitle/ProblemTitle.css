.title-instructions {
  font-size: 4rem;
  padding-top: 13rem;
  color: #f8f8ff; }

p {
  font-size: 1rem;
  padding: 1rem;
  color: #a0a0a0; }

input {
  background-color: rgba(0, 0, 0, 0);
  color: #ffaaaa;
  border: none;
  margin: 4rem auto;
  display: block;
  width: 54rem;
  font-size: 2rem;
  height: 4rem;
  text-align: center; }
  input::placeholder {
    color: #ffaaaa91; }
  input:focus {
    outline: none; }

img {
  width: 1rem; }
