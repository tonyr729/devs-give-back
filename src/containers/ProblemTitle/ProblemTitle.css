.welcome {
  font-size: 2rem;
  padding-top: 10rem; }

.title-instructions {
  font-size: 4rem;
  padding-top: 4rem;
  color: #f8f8ff; }

.example {
  font-size: 1rem;
  padding: 1rem;
  color: #a0a0a0; }

.title-input {
  background-color: rgba(0, 0, 0, 0);
  color: #ffaaaa;
  border: none;
  border-bottom: 1px solid #ffaaaa;
  margin: 4rem auto;
  display: block;
  width: 58rem;
  font-size: 2rem;
  height: 4rem;
  text-align: center; }
  .title-input::placeholder {
    color: #ff5990; }
  .title-input:focus {
    border-bottom: none;
    outline: none; }

.title-next-button {
  width: 6rem;
  height: 6rem;
  background: none;
  border: none;
  display: block;
  margin: 10.5rem auto 0; }
  .title-next-button img {
    width: 4rem; }
  .title-next-button:hover {
    color: #03a9f4; }
