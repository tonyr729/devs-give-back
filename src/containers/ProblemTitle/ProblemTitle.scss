.welcome {
  font-size: 2rem;
  padding-top: 10rem;
}

.title-instructions {
  font-size: 4rem;
  padding-top: 4rem;
  color: #f8f8ff
}

.example {
  font-size: 1rem;
  padding: 1rem;
  color: #a0a0a0;
}

input {
  background-color: rgba(0, 0, 0, 0);
  color: #ffaaaa;
  border: none;
  margin: 4rem auto;
  display: block;
  width: 54rem;
  font-size: 2rem;
  height: 4rem;
  text-align: center;
  &::placeholder {
    color: #ffaaaa91;
  }
  &:focus {
    outline: none;
  }
}

.next-button {
  width: 6rem;
  height: 6rem;
  background: none;
  border: none;
  display: block;
  margin: 12rem auto;
  img {
    width: 4rem;
  }
}
