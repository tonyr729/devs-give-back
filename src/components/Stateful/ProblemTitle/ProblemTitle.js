import React, { Component } from 'react';
import './ProblemTitle.css';
import { NavLink } from 'react-router-dom'
import { connect } from 'react-redux';
import {createProblemTitle, createProblemClient } from '../../../actions/actions';
import rightArrow from '../../../images/right_arrow.svg'

class ProblemTitle extends Component {
  constructor() {
    super();
    this.state = {
      input: ''
    }
  }

  handleInputChange= (event) => {
    this.setState({
      input: event.target.value
    })
  }

  handleSubmit = (event) => {
    event.preventDefault();
    this.props.createProblemTitle(this.state.input);
    this.props.createProblemClient(this.props.user.id);
    this.props.history.push("/problem-body")
  }

  render() {
    let name;
    !this.props.user.name ? this.props.history.push("/client-login") : name = this.props.user.name.split(' ')[0];

    return (
      <div>
        <p className='welcome'>Hi {name},</p>
        <p className='title-instructions'>Briefly describe your problem...</p>
        <p className='example'>Ex: Looking for a solution to help an afterschool program</p> 
        <input className='title-input' placeholder='Looking for ...' maxLength='70' value={this.state.input} onChange={this.handleInputChange}/>
        <button className='next-button' onClick={this.handleSubmit} ><img className='right-arrow-img' src={rightArrow} /></button>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  user: state.user
});

const mapDispatchToProps = (dispatch) => ({
  createProblemClient: (user) => dispatch(createProblemClient(user)),
  createProblemTitle: (title) => dispatch(createProblemTitle(title))
});


export default connect(mapStateToProps, mapDispatchToProps)(ProblemTitle);