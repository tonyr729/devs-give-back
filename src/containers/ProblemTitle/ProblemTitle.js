import React, { Component } from 'react';
import './ProblemTitle.css';
import { NavLink } from 'react-router-dom'
import { connect } from 'react-redux';
import {createProblemTitle, createProblemClient } from '../../actions/actions';
import DatabaseHelper from '../../helpers/DatabaseHelper';
import rightArrow from '../../images/right_arrow.svg';

class ProblemTitle extends Component {
  constructor() {
    super();
    this.state = {
      input: ''
    }

    this.database = new DatabaseHelper();
  }

  componentDidMount(){
    this.problemCheck(this.props.client)
    this.logInCheck(this.props.client);
  }

  problemCheck = async (client) =>{
    client.name || this.props.history.push("/client-login")
    const existingProblem = await this.database.findMatchingProblem(client.id);
    !existingProblem || this.props.history.push("/prior-problem");
  }

  handleInputChange= (event) => {
    this.setState({
      input: event.target.value
    })
  }

  handleSubmit = (event) => {
    event.preventDefault();
    this.props.createProblemTitle(this.state.input);
    this.props.createProblemClient(this.props.client.id);
    this.props.history.push("/problem-body")
  }

  logInCheck = (client) => {
    const value = Object.keys(client).length;
    if ( value === 0 ) {
      this.props.history.push('/client-login')
    }
  }


  render() {
    const name = this.props.client.name || 'client unknown';
    const firstName = name.split(' ')[0];
    return (
      <div className='client-background'>
        <p className='welcome'>Hi {firstName},</p>
        <p className='title-instructions'>Briefly describe your problem...</p>
        <p className='example'>Ex: Looking for a solution to help an afterschool program</p> 
        <input className='title-input' placeholder='Looking for ...' maxLength='70' value={this.state.input} onChange={this.handleInputChange}/>
        <button className='title-next-button' onClick={this.handleSubmit} ><img className='right-arrow-img' src={rightArrow} /></button>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  client: state.client
});

const mapDispatchToProps = (dispatch) => ({
  createProblemClient: (client) => dispatch(createProblemClient(client)),
  createProblemTitle: (title) => dispatch(createProblemTitle(title))
});


export default connect(mapStateToProps, mapDispatchToProps)(ProblemTitle);