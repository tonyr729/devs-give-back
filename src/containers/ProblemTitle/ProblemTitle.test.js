import { ProblemTitle, mapStateToProps, mapDispatchToProps} from './ProblemTitle';
import React from 'react';
import { shallow } from 'enzyme';
jest.mock('../../helpers/DatabaseHelper.js', () => {
  return jest.fn().mockImplementation(() => {
    return {
      findMatchingProblem: jest.fn().mockImplementation(() => Promise.resolve({category: 'category'})),
      writeProblemToDatabase: jest.fn().mockImplementation(() => Promise.resolve({problem: 'problem'}))
    };
  });
});

describe('ProblemTitle', () => {
  it('should match snapshot', () => {
    const mockProps = {
      client: {name: 'tony'},
      history: {
        push: jest.fn()
      }
    };
    const wrapper = shallow(<ProblemTitle {...mockProps} />);

    expect(wrapper).toMatchSnapshot();
  });

  describe('componentDidMount', () => {
    let mockProps;
    let wrapper;

    beforeEach(() => {
      mockProps = {
        client: {name: 'tony'},
        history: {
          push: jest.fn()
        }
      };
      wrapper = shallow(<ProblemTitle {...mockProps} />)
      wrapper.instance().logInCheck = jest.fn();
    });
    
    it('should call given functions on mount of component', async () => {
      // wrapper.instance().problemCheck = jest.fn().mockImplementation(() => Promise.resolve({category: 'category'}));
      // expect(await wrapper.instance().problemCheck).toHaveBeenCalled();
      // expect(wrapper.instance().logInCheck).toHaveBeenCalled();
    });
  });

  describe('problemCheck', () => {
    let mockProps;
    let wrapper;

    beforeEach(() => {
      mockProps = {
        client: {name: 'tony'},
        history: {
          push: jest.fn()
        }
      };
      wrapper = shallow(<ProblemTitle {...mockProps} />)
    });
    
    it('should redirect if no client', async () => {
      wrapper.instance().problemCheck({})

      expect(wrapper.instance().props.history.push).toHaveBeenCalledWith('/client-login')
    });

    it('should call database method', async () => {
      wrapper.instance().problemCheck({name: 'tony'})

      expect(wrapper.instance().database.findMatchingProblem).toHaveBeenCalled()
    });

    it('should redirect if no existing problem', async () => {
      wrapper.instance().problemCheck({name: 'tony'})

      expect(wrapper.instance().props.history.push).toHaveBeenCalledWith('/prior-problem')
    });
  });

  describe('handleInputChange', () => {
    let mockProps;
    let mockEvent;
    let wrapper;

    beforeEach(() => {
      mockProps = {
        client: {name: 'tony'},
        history: {
          push: jest.fn()
        }
      };
      mockEvent = {
        target: {
          value: 'input value'
        }
      }
      wrapper = shallow(<ProblemTitle {...mockProps} />);
    });

    it('should setState off the params', () => {
      wrapper.instance().handleInputChange(mockEvent)
    
      expect(wrapper.state('input')).toEqual(mockEvent.target.value)
    });
  });

  describe('handleInputChange', () => {
    let mockProps;
    let mockEvent;
    let wrapper;

    beforeEach(() => {
      mockProps = {
        client: {name: 'tony'},
        createProblemTitle: jest.fn(),
        createProblemClient: jest.fn(),
        history: {
          push: jest.fn()
        }
      };
      mockEvent = {
        preventDefault: jest.fn(),
        target: {
          value: 'input value'
        }
      }
      wrapper = shallow(<ProblemTitle {...mockProps} />);
    });

    it('should setState off the params', () => {
      wrapper.instance().handleSubmit(mockEvent)
    
      expect(wrapper.instance().props.createProblemTitle).toHaveBeenCalled();
      expect(wrapper.instance().props.createProblemClient).toHaveBeenCalled();
      expect(wrapper.instance().props.history.push).toHaveBeenCalledWith("/problem-body");
    });
  });

  describe('submitInput', () => {
    let mockProps;
    let mockEvent;
    let wrapper;

    beforeEach(() => {
      mockProps = {
        client: {name: 'tony'},
        createProblemTitle: jest.fn(),
        history: {
          push: jest.fn()
        }
      };
      mockEvent = {
        keyCode: 13,
        shiftKey: false,
        preventDefault: jest.fn()
      };
      wrapper = shallow(<ProblemTitle {...mockProps} />);
    });

    it('should call the given methods if keyCodes match', () => {
      wrapper.instance().handleSubmit = jest.fn();
      
      wrapper.instance().submitInput(mockEvent);
    
      expect(wrapper.instance().handleSubmit).toHaveBeenCalled();
    });

    it('should do nothing if keyCodes dont match', () => {
      mockEvent = {
        keyCode: 11,
        shiftKey: true,
        preventDefault: jest.fn()
      };
      wrapper.instance().handleSubmit = jest.fn();

      wrapper.instance().submitInput(mockEvent);
    
      expect(wrapper.instance().handleSubmit).not.toHaveBeenCalled();
    });
  });

  describe('mapStateToProps', () => {
    it('returns an object with client info', () => {

      const mockState = {
        client: 'value',
      }

      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(mockState);
      
    })
  })

  describe('mapDispatchToProps', () => {
    it('should call dispatch on createProblemClient with the correct params', () => {

      const mockDispatch = jest.fn();
      const client = {prob: 'prob'}
      const mappedProps = mapDispatchToProps(mockDispatch);
      const mockAction = {
        type: 'CREATE_PROBLEM_CLIENT',
        client
      };
      
      mappedProps.createProblemClient(client);
      expect(mockDispatch).toHaveBeenCalledWith(mockAction);
    });

    it('should call dispatch on createProblemTitle with the correct params', () => {

      const mockDispatch = jest.fn();
      const title = {prob: 'prob'}
      const mappedProps = mapDispatchToProps(mockDispatch);
      const mockAction = {
        type: 'CREATE_PROBLEM_TITLE',
        title
      };
      
      mappedProps.createProblemTitle(title);
      expect(mockDispatch).toHaveBeenCalledWith(mockAction);
    });
  });
});