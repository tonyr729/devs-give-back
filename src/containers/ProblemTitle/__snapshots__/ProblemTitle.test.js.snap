// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProblemTitle should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ProblemTitle
    client={
        Object {
            "name": "tony",
          }
    }
    history={
        Object {
            "push": [Function],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <p
          className="welcome"
>
          Hi 
          tony
          ,
</p>,
        <p
          className="title-instructions"
>
          Briefly describe your problem...
</p>,
        <p
          className="example"
>
          Ex: Looking for a solution to help an afterschool program
</p>,
        <input
          autoFocus={true}
          className="title-input"
          maxLength="70"
          onChange={[Function]}
          onKeyDown={[Function]}
          value=""
/>,
        <button
          className="title-next-button"
          onClick={[Function]}
>
          <img
                    className="right-arrow-img"
                    src="right_arrow.svg"
          />
</button>,
      ],
      "className": "client-background",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Hi ",
            "tony",
            ",",
          ],
          "className": "welcome",
        },
        "ref": null,
        "rendered": Array [
          "Hi ",
          "tony",
          ",",
        ],
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Briefly describe your problem...",
          "className": "title-instructions",
        },
        "ref": null,
        "rendered": "Briefly describe your problem...",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Ex: Looking for a solution to help an afterschool program",
          "className": "example",
        },
        "ref": null,
        "rendered": "Ex: Looking for a solution to help an afterschool program",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "autoFocus": true,
          "className": "title-input",
          "maxLength": "70",
          "onChange": [Function],
          "onKeyDown": [Function],
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <img
            className="right-arrow-img"
            src="right_arrow.svg"
/>,
          "className": "title-next-button",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "right-arrow-img",
            "src": "right_arrow.svg",
          },
          "ref": null,
          "rendered": null,
          "type": "img",
        },
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <p
            className="welcome"
>
            Hi 
            tony
            ,
</p>,
          <p
            className="title-instructions"
>
            Briefly describe your problem...
</p>,
          <p
            className="example"
>
            Ex: Looking for a solution to help an afterschool program
</p>,
          <input
            autoFocus={true}
            className="title-input"
            maxLength="70"
            onChange={[Function]}
            onKeyDown={[Function]}
            value=""
/>,
          <button
            className="title-next-button"
            onClick={[Function]}
>
            <img
                        className="right-arrow-img"
                        src="right_arrow.svg"
            />
</button>,
        ],
        "className": "client-background",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Hi ",
              "tony",
              ",",
            ],
            "className": "welcome",
          },
          "ref": null,
          "rendered": Array [
            "Hi ",
            "tony",
            ",",
          ],
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Briefly describe your problem...",
            "className": "title-instructions",
          },
          "ref": null,
          "rendered": "Briefly describe your problem...",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Ex: Looking for a solution to help an afterschool program",
            "className": "example",
          },
          "ref": null,
          "rendered": "Ex: Looking for a solution to help an afterschool program",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "autoFocus": true,
            "className": "title-input",
            "maxLength": "70",
            "onChange": [Function],
            "onKeyDown": [Function],
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <img
              className="right-arrow-img"
              src="right_arrow.svg"
/>,
            "className": "title-next-button",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "right-arrow-img",
              "src": "right_arrow.svg",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
